go/consensus/tendermint/apps/roothash/slashing.go

```diff
@@ -0,0 +1,89 @@
+package roothash
+
+import (
+	"fmt"
+
+	"github.com/oasisprotocol/oasis-core/go/common"
+	"github.com/oasisprotocol/oasis-core/go/common/cbor"
+	"github.com/oasisprotocol/oasis-core/go/common/crypto/signature"
+	"github.com/oasisprotocol/oasis-core/go/common/quantity"
+	"github.com/oasisprotocol/oasis-core/go/consensus/tendermint/api"
+	abciAPI "github.com/oasisprotocol/oasis-core/go/consensus/tendermint/api"
+	registryState "github.com/oasisprotocol/oasis-core/go/consensus/tendermint/apps/registry/state"
+	stakingState "github.com/oasisprotocol/oasis-core/go/consensus/tendermint/apps/staking/state"
+	roothash "github.com/oasisprotocol/oasis-core/go/roothash/api"
+	staking "github.com/oasisprotocol/oasis-core/go/staking/api"
+)
+
+func onEvidenceRuntimeEquivocation(
+	ctx *abciAPI.Context,
+	pk signature.PublicKey,
+	runtimeID common.Namespace,
+) error {
+	regState := registryState.NewMutableState(ctx.State())
+	stakeState := stakingState.NewMutableState(ctx.State())
+
+	node, err := regState.Node(ctx, pk)
+	if err != nil {
+		// Node might not exist anymore (old evidence). Or the submitted evidence
+		// could be for a non-existing node (submitting "fake" but valid evidence).
+		ctx.Logger().Error("failed to get runtime node by signature public key",
+			"public_key", pk,
+			"err", err,
+		)
+		// TODO: Should this be treated as an error?
+		return fmt.Errorf("node not found: %w", roothash.ErrInvalidEvidence)
+	}
+
+	// Retrieve the slash procedure for double signing.
+	st, err := stakeState.Slashing(ctx)
```

Comment:
This should use the per-runtime slashing procedure.

Reply:
Yeah of course, thanks! Will also update the transactions test to ensure the expected amount was slashed (which would've found this).